src/Authoring/WinRT.SourceGenerator/AotOptimizer.cs

```diff
@@ -0,0 +1,210 @@
+﻿using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+
+namespace Generator
+{
+    [Generator]
+    public class WinRTAotSourceGenerator : IIncrementalGenerator
+    {
+        public void Initialize(IncrementalGeneratorInitializationContext context)
+        {
+            var vtableAttributesToAdd = context.SyntaxProvider.CreateSyntaxProvider(
+                static (n, _) => NeedVtableAttribute(n),
+                static (n, _) => GetVtableAttributeToAdd(n));
+
+            context.RegisterImplementationSourceOutput(vtableAttributesToAdd, GenerateVtableAttributes);
+        }
+
+        // Restrict to non-projected classes which can be instantiated
+        // and are partial allowing to add attributes.
+        private static bool NeedVtableAttribute(SyntaxNode node)
+        {
+            return node is ClassDeclarationSyntax declaration &&

```

Comment:
@jkoritzinsky do you think using a syntax provider with this filtering is enough to not make the transform resolving symbols be too expensive? I know that at least for now, `RegisterImplementationSourceOutput` isn't yet wired to actually skip work during normal edits, so just wondering whether there might be any concerns here 🤔